internal/providers/github/github_rest.go

```diff
@@ -609,10 +610,30 @@ func (c *RestClient) ListPullRequests(
 
 // CreateComment creates a comment on a pull request or an issue
 func (c *RestClient) CreateComment(ctx context.Context, owner, repo string, number int, comment string) error {
-	_, _, err := c.client.Issues.CreateComment(ctx, owner, repo, number, &github.IssueComment{
-		Body: &comment,
-	})
-	return err
+	op := func() error {
+		// Check if the token is currently rate limited, block the goroutine if it is
+		err := c.waitIfRateLimited(ctx)
+		if err != nil {
+			return err
+		}
+
+		_, resp, err := c.client.Issues.CreateComment(ctx, owner, repo, number, &github.IssueComment{
+			Body: &comment,
+		})
+
+		rateLimCheckErr := c.checkRateLimitResponse(ctx, resp.Response)
+		if rateLimCheckErr != nil {
+			return rateLimCheckErr
+		}
+
+		// If err is of type RateLimitError or AbuseRateLimitError, don't do permanent backoff
+		if isRateLimitError(err) {
+			return err
+		}
+
+		return backoffv4.Permanent(err)
+	}
+	return performWithRetry(ctx, op)
 }
```

Comment:
`CreateComment` endpoint is modified as an example. If we agree on the implementation, other endpoints will be modified as well.

Reply:
I like this pattern, but I wonder if we can make this more generic to avoid repeating the closure around all GH calls. Taking your `CreateComment` example into account, could we do something like:
```go
func (c *RestClient) CreateComment(ctx context.Context, owner, repo string, number int, comment string) error {
    createCommentWrapper := func(ctx context.Context) (*github.Response, error) {
        return c.client.Issues.CreateComment(ctx, owner, repo, number, &github.IssueComment{
            Body: &comment,
        })
    }

    return c.rateLimitHandler(ctx, createCommentWrapper)
}
```

and then extract the closure into a non-exported function:
```go
type ghClientWrapper func(ctx context.Context) (*github.Response, error)

func (c *RestClient) rateLimitHandler(ctx context.Context, wrapper ghClientWrapper) error {
    op := func() error {
        // Check if the token is currently rate limited, block the goroutine if it is
        err := c.waitIfRateLimited(ctx)
        if err != nil {
            return err
        }

        _, resp, err := apiCall(ctx)
        if err != nil {
            rateLimCheckErr := c.checkRateLimitResponse(ctx, resp.Response)
            if rateLimCheckErr != nil {
                return rateLimCheckErr
            }

            // If err is of type RateLimitError or AbuseRateLimitError, don't do permanent backoff
            if isRateLimitError(err) {
                return err
            }

            return backoffv4.Permanent(err)
        }

        return nil
    }

    return performWithRetry(ctx, op)
}
```